custom_components/foxess_modbus/entities/modbus_time_period_sensors.py

```diff
@@ -0,0 +1,138 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.helpers.restore_state import ExtraStoredData
+from homeassistant.helpers.restore_state import RestoredExtraData
+from homeassistant.helpers.restore_state import RestoreEntity
+
+from ..common.entity_controller import EntityController
+from .modbus_entity_mixin import ModbusEntityMixin
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(kw_only=True)
+class ModbusTimePeriodStartEndSensorDescription(SensorEntityDescription):
+    """Entity description for ModbusTimePeriodStartEndSensor"""
+
+    address: int
+    other_address: int  # Address of period end if this is the start, and vice versa
+
+
+class ModbusTimePeriodStartEndSensor(ModbusEntityMixin, RestoreEntity, SensorEntity):
+    """Sensor used for the start/end of a charge time period"""
+
+    def __init__(
+        self,
+        controller: EntityController,
+        entity_description: ModbusTimePeriodStartEndSensorDescription,
+        entry: ConfigEntry,
+        inv_details,
+    ) -> None:
+        self._controller = controller
+        self.entity_description = entity_description
+        self._entry = entry
+        self._inv_details = inv_details
+        self.entity_id = "sensor." + self._get_unique_id()
+        self._last_non_zero_value: int | None = None
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        value = self._controller.read(self.entity_description.address)
+        if value is not None:
+            # If the charge window is disabled (i.e. both start and end are 0),
+            # return the last-stored value rather than midnight
+            if value == 0 and self._last_non_zero_value is not None:
+                other_value = self._controller.read(
+                    self.entity_description.other_address
+                )
+                if other_value == 0:
+                    value = self._last_non_zero_value
+
+            value = time(hour=(value & 0xFF00) >> 8, minute=value & 0xFF)
+
+        return value
+
+    async def async_added_to_hass(self) -> None:
+        """Add update callback after being added to hass."""
+        await super().async_added_to_hass()
+        extra_data = await self.async_get_last_extra_data()
+        if extra_data:
+            self._last_non_zero_value = extra_data.json_dict.get("last_non_zero_value")
+
+    @property
+    def extra_restore_state_data(self) -> ExtraStoredData:
+        """Return specific state data to be restored."""
+        return RestoredExtraData(
+            json_dict={"last_non_zero_value": self._last_non_zero_value}
+        )
+
+    def update_callback(self, changed_addresses: set[int]) -> None:
+        """Schedule a state update."""
+        if self.entity_description.address in changed_addresses:
+            value = self._controller.read(self.entity_description.address)
+            if value is not None and value > 0:
```

Comment:
This isn't quite right, as it's valid to set a start time of midnight. We need to check the other sensor as well.